### PR TITLE
LUN-100: Implement article cms changes

### PR DESCRIPTION
**Pull request by SteffiStuffel**
This pull request belongs to an issue on our [bugtracker](https://issues.tuerantuer.org/secure/RapidBoard.jspa?rapidView=27&view=detail).
You can find it there by looking for an issue with the key which is mentioned in the title of this pull request.
It starts with the keyword _LUN_.

### CODE REVIEW

File: src/components/VocabularyListItem.tsx [hunk 1]
```diff
@@ -79,17 +78,19 @@ const VocabularyListItem = ({ document }: VocabularyListItemPropType) => {
...
+              }}
+              width={24}
+              height={24}
+              style={styles.image}
+            />
+          )}
           <View>
             <Text testID='article' style={[styles.title, { backgroundColor: getArticleColor(article) }]}>
-              {article?.toLowerCase() === ARTICLES.diePlural ? 'Die' : capitalizeFirstLetter(article)}
+              {article.value}
```
Comment (klinzo) [inline src/components/VocabularyListItem.tsx:93]: Previously this was capitalized:
```suggestion
              {capitalizeFirstLetter(article.value)}
```

File: src/components/VocabularyListItem.tsx [hunk 2]
```diff
@@ -79,17 +78,19 @@ const VocabularyListItem = ({ document }: VocabularyListItemPropType) => {
...
-              uri: document.document_image[0].image
-            }}
-            width={24}
-            height={24}
-            style={styles.image}
-          />
+          {document.document_image.length > 0 && (
+            <Image
+              source={{
+                uri: document?.document_image[0].image
```
Comment (klinzo) [inline src/components/VocabularyListItem.tsx:84]:
```suggestion
                uri: document.document_image[0].image
```
as the check is not used in line 81 as well.

File: src/hooks/useLoadDocuments.ts [hunk 3]
```diff
@@ -1,9 +1,38 @@
...
+      return article.id === doc.article
+    })
+    if (!formattedArticle) {
+      throw new Error('Invalid Server Response')
+    }
+    const formattedDocument: DocumentType = { ...rest, article: formattedArticle }
+    documentsWithTransformedArticle.push(formattedDocument)
+  })
+  const { data, ...rest } = documents
+  return { ...rest, data: documentsWithTransformedArticle }
```
Comment (klinzo) [inline src/hooks/useLoadDocuments.ts:29]: Using `.map()` this becomes simpler :)

File: src/screens/choice-exercises/components/SingleChoiceExercise.tsx [hunk 4]
```diff
@@ -49,32 +49,20 @@ const ChoiceExerciseScreen = ({ documents, documentToAnswers, onExerciseFinished
...
-        found = true
-      }
-    })
-    return found
+    return answer1.article.id === answer2.article.id && answer1.word === answer2.word
   }
 
   const correctAlternatives: Answer[] = currentDocument.alternatives.map(it => ({
-    article: it.article,
+    article: ARTICLES[it.article],
```
Comment (klinzo) [inline src/screens/choice-exercises/components/SingleChoiceExercise.tsx:56]: Sometimes you actually compare the id, sometimes you just access the array with the id as index, I guess we should just use one of both.

File: src/screens/choice-exercises/WordChoiceExerciseScreen.tsx [hunk 5]
```diff
@@ -55,11 +55,14 @@ const WordChoiceExerciseScreen = ({ navigation, route }: WordChoiceExerciseScree
   }
 
   return (
-    <SingleChoiceExercise
-      documents={documents}
-      documentToAnswers={documentToAnswers}
-      onExerciseFinished={onExerciseFinished}
-    />
+    documents !== null &&
+    documents.length !== 0 && (
```
Comment (klinzo) [inline src/screens/choice-exercises/WordChoiceExerciseScreen.tsx:59]: Why is this needed?

File: src/screens/WriteExerciseScreen.tsx [hunk 6]
```diff
@@ -79,29 +78,14 @@ const WriteExerciseScreen = ({ navigation, route }: WriteExerciseScreenPropsType
...
-      getDocuments()
-    }, [retryData, trainingSetId])
-  )
+  const [currentDocumentNumber, setCurrentDocumentNumber] = useState(0)
+
+  let documents = useLoadDocuments(trainingSetId).data
+
+  if (retryData) {
+    documents = retryData.data
+  }
```
Comment (klinzo) [inline src/screens/WriteExerciseScreen.tsx:88]:
```suggestion
  const documents = retryData?.data ?? useLoadDocuments(trainingSetId).data
```

File: src/screens/WriteExerciseScreen.tsx [hunk 7]
```diff
@@ -113,7 +97,7 @@ const WriteExerciseScreen = ({ navigation, route }: WriteExerciseScreenPropsType
           </TouchableOpacity>
         ),
         headerRight: () => (
-          <Text style={styles.headerText}>{`${currentDocumentNumber + 1} of ${documents.length}`}</Text>
+          <Text style={styles.headerText}>{`${currentDocumentNumber + 1} of ${documents?.length}`}</Text>
```
Comment (klinzo) [inline src/screens/WriteExerciseScreen.tsx:100]: Is this displayed if there are no documents as well? If so, it probably says something like `1 of undefined`

File: src/screens/WriteExerciseScreen.tsx [hunk 8]
```diff
@@ -132,18 +116,21 @@ const WriteExerciseScreen = ({ navigation, route }: WriteExerciseScreenPropsType
...
   }
 
   const finishExercise = (): void => {
     AsyncStorage.clearSession().catch(e => console.error(e))
+    setCurrentDocumentNumber(0)
     navigation.navigate('InitialSummary', { extraParams: { ...extraParams, results: [] } })
   }
 
-  const docsLength = documents.length
+  const docsLength = documents ? documents.length : 0
```
Comment (klinzo) [inline src/screens/WriteExerciseScreen.tsx:133]:
```suggestion
  const docsLength = documents?.length ?? 0
```

**Review by klinzo — APPROVED** [review]

File: src/components/VocabularyListItem.tsx (hunk 1 above)
Comment (SteffiStuffel) [reply src/components/VocabularyListItem.tsx:93]: Ich habe jetzt in data.ts einfach mal die ersten Buchstaben groß gemacht, dann können wir uns das hier sparen und ich denke wir brauchen es nie in klein.

File: src/components/VocabularyListItem.tsx (hunk 2 above)
Comment (SteffiStuffel) [reply src/components/VocabularyListItem.tsx:84]: done

File: src/screens/choice-exercises/components/SingleChoiceExercise.tsx (hunk 4 above)
Comment (SteffiStuffel) [reply src/screens/choice-exercises/components/SingleChoiceExercise.tsx:56]: yeah, this is my quickfix for the alternative-words, where the mapping from id to Article is not jet implemented. I created an issue for it.

File: src/screens/choice-exercises/WordChoiceExerciseScreen.tsx (hunk 5 above)
Comment (SteffiStuffel) [reply src/screens/choice-exercises/WordChoiceExerciseScreen.tsx:59]: it is not i guess. removed it.

Comment (SteffiStuffel) [reply src/screens/choice-exercises/WordChoiceExerciseScreen.tsx:59]: ok, it is needed, added it again. If loading is not finished, it needs to block rendering.

File: src/screens/WriteExerciseScreen.tsx (hunk 7 above)
Comment (SteffiStuffel) [reply src/screens/WriteExerciseScreen.tsx:100]: if shows `1 of 0`, happens while loading is not finished, think this behaviour is oke for now.

File: src/screens/WriteExerciseScreen.tsx (hunk 8 above)
Comment (SteffiStuffel) [reply src/screens/WriteExerciseScreen.tsx:133]: done

File: src/screens/WriteExerciseScreen.tsx (hunk 6 above)
Comment (SteffiStuffel) [reply src/screens/WriteExerciseScreen.tsx:88]: Yeah, wanted to do this as well, but it does not work as `ESLint: React Hook "useLoadDocuments" is called conditionally. React Hooks must be called in the exact same order in every component render.(react-hooks/rules-of-hooks`

File: src/hooks/useLoadDocuments.ts (hunk 3 above)
Comment (SteffiStuffel) [reply src/hooks/useLoadDocuments.ts:29]: yeah, strugggling a bit with this, because it changes the types of the object. I will have a closer look at this later (after deployment), as i do not have time and patience now for it.

File: src/screens/choice-exercises/WordChoiceExerciseScreen.tsx (hunk 5 above)
Comment (klinzo) [reply src/screens/choice-exercises/WordChoiceExerciseScreen.tsx:59]: But shouldn't there be a `loading` prop which does exactly that?

Comment (SteffiStuffel) [reply src/screens/choice-exercises/WordChoiceExerciseScreen.tsx:59]: Yeah it should and it is, but if removing i get ` Type 'DocumentType[] | null' is not assignable to type 'DocumentType[]'.   Type 'null' is not assignable to type 'DocumentType[]'.`

Because ReturnType<DocumentType[]> also allows `null`. I don't know if there is a good way to tell typescript that it will only be null if loading is true.

Comment (SteffiStuffel) [reply src/screens/choice-exercises/WordChoiceExerciseScreen.tsx:59]: oke sorry, i was talking about the first line, you are right, the second line can be replaced with `!loading`
